DataFormats/Headers/include/Headers/DataHeader.h

```diff
@@ -460,77 +467,109 @@ auto get(const void* buffer, size_t len = 0)
 ///   - returns a Stack ready to be shipped.
 struct Stack {
- protected:
-  using Buffer = std::unique_ptr<byte[]>;
-  size_t bufferSize;
-  Buffer buffer;
-
- public:
-  // This is ugly and needs fixing BUT:
-  // we need a static deleter for fairmq.
-  // TODO: maybe use allocator_traits if custom allocation is desired
-  //       the deallocate function is then static.
-  //       In the case of special transports is is cheap to copy the header stack into a message, so no problem there.
-  //       The copy can be avoided if we construct in place inside a FairMQMessage directly (instead of
-  //       allocating a unique_ptr we would hold a FairMQMessage which for the large part has similar semantics).
-  //
-  static std::default_delete<byte[]> sDeleter;
-  static void freefn(void* /*data*/, void* hint) {
-    Stack::sDeleter(static_cast<byte*>(hint));
-  }
 
-  byte* data() const {return buffer.get();}
-  size_t size() const {return bufferSize;}
-  void release() {buffer.release();}
-
-  ///The magic constructor: takes arbitrary number of arguments and serialized them
-  /// into the buffer.
-  /// Intended use: produce a temporary via an initializer list.
-  /// TODO: maybe add a static_assert requiring first arg to be DataHeader
-  /// or maybe even require all these to be derived form BaseHeader
-  template<typename... Headers>
-  Stack(const Headers&... headers)
-    : bufferSize{size(headers...)}
-    , buffer{std::make_unique<byte[]>(bufferSize)}
+  static void freefn(void* data, void* hint)
   {
-    inject(buffer.get(), headers...);
+    boost::container::pmr::memory_resource* resource = static_cast<boost::container::pmr::memory_resource*>(hint);
+    resource->deallocate(data, 0, 0);
   }
+
+  struct freeobj {
+    boost::container::pmr::memory_resource* resource{ nullptr };
+    void operator()(byte* ptr) { Stack::freefn(ptr, resource); }
+  };
+
+  using allocator_type = boost::container::pmr::polymorphic_allocator<byte>;
+  using value_type = byte;
+  using BufferType = std::unique_ptr<byte[], freeobj>;
+
   Stack() = default;
   Stack(Stack&&) = default;
   Stack(Stack&) = delete;
   Stack& operator=(Stack&) = delete;
   Stack& operator=(Stack&&) = default;
 
-  template<typename T, typename... Args>
-  static size_t size(const T& h, const Args... args) noexcept {
-    return size(h) + size(args...);
+  byte* data() const { return buffer.get(); }
+  size_t size() const { return bufferSize; }
+  void release() { buffer.release(); }
+  void clear() { buffer.release(); }
+  allocator_type get_allocator() const { return allocator; }
+
+  /// The magic constructors: take arbitrary number of headers and serialize them
+  /// into the buffer buffer allocated by the specified polymorphic allocator. By default
+  /// allocation is done using new_delete_resource.
+  /// The first argument MUST be DataHeader or another Stack
+  /// all others must derive from BaseHeader and not be DataHeader
```

Comment:
is it a hard requirement that there is not more than one `DataHeader` in the Stack? Should have a check for this.

Reply:
I'd say it would not make much sense to have more than one DataHeader in the stack (since all get<> methods just stop searching after first hit, so in a way enforce one single header of each type per stack), but so far I have not been able to figure out a good check for this. I can remove it from the documentation, but maybe it would make sense to keep it in there as a reminder (a convention).